Introduce INDEX_SHARD_COUNT_FORMAT

joshua-adams-1 opened the pull request:
Replaces the existing `ChecksumBlobStoreFormat` `INDEX_METADATA_FORMAT` inside `BlobStoreRepository` with a new `INDEX_SHARD_COUNT_FORMAT`. This removes the need for writing an entire `IndexMetadata` object to, and reading from, heap memory. Instead, we write to and read from heap only the shard count for an index, reducing the total heap memory needed for snapshotting and reducing the likelihood a node will go OOMe.

Relates to https://github.com/elastic/elasticsearch/issues/131822
Relates to ES-12539
Closes ES-12538

joshua-adams-1 committed:
Replace INDEX_METADATA_FORMAT with INDEX_SHARD_COUNT_FORMAT

Replaces the existing ChecksumBlobStoreFormat INDEX_METADATA_FORMAT
inside BlobStoreRepository with a new INDEX_SHARD_COUNT_FORMAT. This
removes the need for writing an entire IndexMetadata object to, and
reading from, heap memory. Instead, we write to and read from heap only
the shard count for an index, reducing the total heap memory needed for
snapshotting and reducing the likelihood a node will go OOMe.

Closes ES-12539
Closes ES-12538

joshua-adams-1 committed:
Remove unused code

joshua-adams-1 committed:
Update docs/changelog/137210.yaml

DaveCTurner reviewed (commented):
> As a note, I have left the existing `INDEX_METADATA_FORMAT` since this is in use elsewhere in the codebase, and can be deprecated afterwards.

I don't think we can deprecate this, no, it's important to be able to read the entire index metadata sometimes (e.g. when restoring an index from snapshot).

joshua-adams-1 committed:
Introduces a new reader to only read the shard count

joshua-adams-1 committed:
Merge branch 'new-index-metadata-format' of https://github.com/joshua-adams-1/elasticsearch into new-index-metadata-format

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -398,6 +398,14 @@ public static String getRepositoryDataBlobName(long repositoryGeneration) {
+    public static final ChecksumBlobStoreFormat<IndexMetadata.IndexShardCount> INDEX_SHARD_COUNT_FORMAT = new ChecksumBlobStoreFormat<>(
+        "index-metadata",
+        METADATA_NAME_FORMAT,
+        (repoName, parser) -> IndexMetadata.IndexShardCount.fromLegacyIndexMetaData(parser),
```
Interesting, TIL, I didn't know we had this notion of a "fallback" parser. Apparently we added it in #78244 to add back support for repositories dating back to 5.0. Any idea what's changed? I'm not sure we need this just to get hold of the settings.

joshua-adams-1 replied:
It was originally to preserve functionality with the existing `INDEX_METADATA_FORMAT`. Since I was still technically parsing the `IndexMetadata` (even if just for one field) I assumed I would also need a legacy and non-legacy parsing method. However, since I am only parsing the shard count and ignoring everything else, there is no difference between versions and this can be removed. Good catch!

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1904,7 +1913,8 @@ record RootBlobUpdateResult(RepositoryData oldRepositoryData, RepositoryData new
 
-                    // Write the index metadata for each index in the snapshot
+                    // Write the index metadata and the shard count to memory for each index in the snapshot, so that it persists
+                    // even if the repository is deleted
```
"write to memory" is kinda odd phrasing - writes are supposed to be persistent here. Maybe this is leftover from an earlier change?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(KEY_SHARD_COUNT, shardCount);
+            return builder;
+        }
```
I don't think we should have a `toXContent` here, we're only using this to read the shard count, not to write it out again

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
+        }
+
+        public static class Builder {
+            private int count;
```
This also seems kinda excessive - we don't need a builder for a class with just one `int` field.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataTests.java:
```diff
@@ -698,6 +604,179 @@ public void testReshardingBWCSerialization() throws IOException {
+        assertEquals(numberOfShards, count.getCount());
+    }
+
+    public void testFromIndexMetaDataWithoutNumberOfShardsSettingReturnsNegativeOne() throws IOException {
```
For these legacy versions, I'd rather we had an actual copy of the index metadata blob written by the older version and verified that we read it as expected. Otherwise if we make a change to `IndexMetadata` serialization and a parallel change to the `IndexShardCount` deserialization then we may not notice a problem with deserializing actual blobs until it's too late. I mean it's ok to synthesize the older versions like this too, but using a real copy of some ancient metadata would be a valuable addition.

DaveCTurner replied:
Ah that's a good point, we could add something to `org.elasticsearch.upgrades.MultiVersionRepositoryAccessIT` to check that we can load the shard count from the blob actually written by the old cluster. We almost cover this already since e.g. `testUpgradeMovesRepoToNewMetaVersion` uses the current version to delete snapshots created by a much older version, except that I think we delete _all_ the snapshots for the old index in one go so we never actually look at its shard count.

However those tests only go back as far as 8.0.0 AFAICT. That's actually a testing gap, I think we need to be able to deal with snapshot repositories going back as far as 5.0.0 (via the `old-lucene-versions` plugin for pre-8.0.0 indices).

DaveCTurner reviewed (commented):
Would it be possible to isolate the changes with `o.e.repositories.blobstore` rather than also adding these things to `IndexMetadata`? This change is very blobstore-specific, I'd rather it wasn't so visible to non-blobstore users of `IndexMetadata`.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
+     */
+    public static class IndexShardCount implements ToXContentFragment {
+        private static final String KEY_SHARD_COUNT = "shard_count";
+        private final int shardCount;
```
I'd prefer all these changes live in blobstore package instead here. It is kinda confusing for people who does not care about how snapshot works.

joshua-adams-1 replied:
I agree, will move them

ywangd commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
+                token = parser.nextToken();
+                if (token == XContentParser.Token.START_OBJECT) {
+                    if (currentFieldName.equals(KEY_SETTINGS)) {
+                        Settings settings = Settings.fromXContent(parser);
```
Since we still need to read the settings, I wonder whether it is necessary to have this as a separate format. The following "inclusive filtering parser" seems to work for me

```java
try (
    XContentParser parser = createParser(
        XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE)
            .withFiltering(Set.of("*.settings", "*.mapping_version", "*.settings_version"), null, false),
        JsonXContent.jsonXContent,
        BytesReference.bytes(builder)
    )
) {
    final var indexMetadata = IndexMetadata.fromXContent(parser);
}
```
It builds an `IndexMetadata` with the minimal number of required information which is not much besides `settings`.

joshua-adams-1 replied:
I have a WIP PR https://github.com/elastic/elasticsearch/pull/137335 if this matches your expectations. If not, I am also working on a new revision of this PR with the above requested changes

ywangd replied:
> if it is still loading some IndexMetadata object with N size, since at this time I don't know the value of N, we may still need to implement concurrency if having 10 threads loading this in parallel would cause a small node to OOMe.

It should just load what is specified in the config, which is `settings` and a few `long`s. Since the new Format reads `settings` as well, I think the memory footprint should be quite close.

> make it hard to add invariants in future

This is true. I was actually a little surprised on how little invariants we have today for the almost empty IndexMetadata.

> I have a WIP PR https://github.com/elastic/elasticsearch/pull/137335 if this matches your expectations

Thanks for testing it out. I personally quite like its simplicity. But I am ok with a new reader as well. Also the latest trimmed down version looks simpler.

joshua-adams-1 replied:
The `ChecksumBlobStoreFormat` deserialization is [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java#L134). The parser is defined [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java#L151) as:
```
XContentParser parser = XContentHelper.createParserNotCompressed(
    XContentParserConfiguration.EMPTY.withRegistry(namedXContentRegistry)
        .withDeprecationHandler(LoggingDeprecationHandler.INSTANCE),
    bytesReference,
    XContentType.SMILE
)
```
There is currently no way to configure this. I extended this method in https://github.com/elastic/elasticsearch/pull/137335 to have a `XContentParserConfiguration` passed in as a parameter, but I think it would be redundant to integrate this change into this PR (as opposed to just having that by itself anyways). If I was defining the parser using a `XContentParserConfiguration` to skip over all the fields other than shard count, would the reader not just become a wrapper to invoking the parser and returning the result?

joshua-adams-1 commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
+                        Settings settings = Settings.fromXContent(parser);
+                        if (legacy) {
+                            checkSettingIndexVersionCompatibility(settings);
+                        }
```
As per my response to [this](https://github.com/elastic/elasticsearch/pull/137210/files#r2470861059) comment, I can remove this

<!--
Currently, the code:
1. Writes an `IndexMetadata` object to the heap
2. Reads the entire `IndexMetadata` object from heap:
2.1 It attempts to parse it using the `legacyFromXContent` method, [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L2865)
2.2 If this fails, it then uses the `fromXContent` method [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L2702)

I want to:
1. Write an `IndexMetadata` object to the heap
2. Read only the shard count from heap

However, to preserve functionality, I opted to follow a similar approach to the above, where I try to parse using the legacy format, and then if this fails, use a newer parser. Having realised two separate parsing functions was redundant, I combined them. 

I am happy to remove this check if you think it too strict, but we would then have this possible situation:

Old:
1. The code fails leha
2. -->

joshua-adams-1 committed:
Comments

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

joshua-adams-1 committed:
Fix tests

joshua-adams-1 committed:
Cleanup

joshua-adams-1 committed:
Minor tweaks

joshua-adams-1 committed:
Merge branch 'new-index-metadata-format' of https://github.com/joshua-adams-1/elasticsearch into new-index-metadata-format

ywangd commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,71 @@
+                } else {
+                    // Iterate through the object, but we don't care for it's contents
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    }
```
It should be performant to call `parser.skipChildren()` and avoid the while loops.

joshua-adams-1 committed:
Adds testSnapshotCreatedInOldVersionCanBeDeletedInNew test

joshua-adams-1 committed:
Update changelog file

joshua-adams-1 committed:
Comments

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

joshua-adams-1 committed:
Revert IndexMetadataTests change that appeared after merging main

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,69 @@
+            if (token == XContentParser.Token.START_OBJECT) {
+                if (currentFieldName.equals(KEY_SETTINGS)) {
+                    Settings settings = Settings.fromXContent(parser);
+                    indexShardCount = new IndexShardCount(settings.getAsInt(SETTING_NUMBER_OF_SHARDS, -1));
```
Could we push the skipping behaviour down even further here? There's no need to load a whole `Settings`. It's not as big as the mappings but can still be sizeable (e.g. `index.query.default_field` could list thousands of field names).

DaveCTurner reviewed (commented):
I like it, much nicer than inventing some arbitrary concurrency limits. I left some comments.

joshua-adams-1 commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,69 @@
+        XContentParser.Token token = parser.nextToken();
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser);
+
+        IndexShardCount indexShardCount = new IndexShardCount(-1);
```
Unfortunately we do need to consume the parser to the end in order to satisfy this constraint, [here](https://github.com/elastic/elasticsearch/blob/c527bbf7e47d9da69e67fda46d857f923deb0a48/server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java#L159):
```
result = reader.apply(projectRepo, parser);
XContentParserUtils.ensureExpectedToken(null, parser.nextToken(), parser);
```

My original solution returned early as an optimisation and threw an error for this reason.

However, I can instantiate the variable to null if you believe that to be better. By setting it to -1 I was protecting us from returning a null in the case of a malformed IndexMetadata object. If this now does occur, this code here:
```
try {
    updateShardCount(
        INDEX_SHARD_COUNT_FORMAT.read(getProjectRepo(), indexContainer, indexMetaGeneration, namedXContentRegistry).count()
    );
} catch (Exception ex) {
    logger.warn(() -> format("[%s] [%s] failed to read shard count for index", indexMetaGeneration, indexId.getName()), ex);
    // Definitely indicates something fairly badly wrong with the repo, but not immediately fatal here: we might get the
    // shard count from another metadata blob, or we might just not process these shards. If we skip these shards
    // then the repository will technically enter an invalid state (these shards' index-XXX blobs will refer to snapshots
    // that no longer exist) and may contain dangling blobs too. A subsequent delete that hits this index may repair
    // the state if the metadata read error is transient, but if not then the stale indices cleanup will eventually
    // remove this index and all its extra data anyway.
    // TODO: Should we fail the delete here? See https://github.com/elastic/elasticsearch/issues/100569.
}
```
will throw a NPE since null.count() isn't possible. The error will be caught and ignored, and we will move on to the next blob. I am happy with this behaviour, just wanted it documented

ywangd replied:
Technically we don't really need the `IndexShardCount` class at all. What we need is an `Integer` and we can make the parser method to return that directly, e.g.:

```java
public static Integer fromIndexMetadata(XContentParser parser) {
    int count = -1;
    // parse for count
    return count;
}

public static final ChecksumBlobStoreFormat<Integer> INDEX_SHARD_COUNT_FORMAT = new ChecksumBlobStoreFormat<>(..);
```

DaveCTurner replied:
Ah yes thanks I'd forgotten that. In principle we could have a similar cache I suppose, particularly for `1`, tho in practice I don't think the extra complexity is worth it. I'd rather keep the clarity of a separate class for now, it's already orders of magnitude less memory usage, and in the (IMO very unlikely) event that we need to save another allocation then we can do so here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -398,6 +398,19 @@ public static String getRepositoryDataBlobName(long repositoryGeneration) {
+     * Parses only the shard count from the IndexMetadata object written by INDEX_METADATA_FORMAT (#131822)
+     */
+    public static final ChecksumBlobStoreFormat<IndexShardCount> INDEX_SHARD_COUNT_FORMAT = new ChecksumBlobStoreFormat<>(
+        "index-metadata",
```
Ah ok gotcha, I missed that we check this there.

joshua-adams-1 committed:
Comments

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

ywangd commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,71 @@
+                        String fieldName = parser.currentName();
+                        parser.nextToken();
+                        if (SETTING_NUMBER_OF_SHARDS.equals(fieldName)) {
+                            indexShardCount = new IndexShardCount(parser.intValue());
```
Nit: can we assert `indexShardCount` is `null` before the assignment?

ywangd reviewed (commented):
This looks good to me. I have a question on whether we want to futher simplify and directly use an Integer in place of the IndexShardCount object.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,71 @@
+            if (token == XContentParser.Token.START_OBJECT) {
+                if (currentFieldName.equals(KEY_SETTINGS)) {
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        String fieldName = parser.currentName();
```
Could we have a little more error-checking here? Technically it's possible to de/serialize a `Settings` as a tree of nested objects (splitting keys at the `.` character) rather than a flat object as we assume here. AFAICT in snapshot context we've always written this as a flat map (going back to 2.4 anyway) but there is a risk that we somehow lose that in future and I'd want that to cause a noisy failure here rather than to just silently skip the child objects.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/IndexShardCount.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * A subset of {@link IndexMetadata} storing only the shard count of an index
+ * Prior to v9.3, the entire {@link IndexMetadata} object was stored in heap and then loaded during snapshotting to determine
+ * the shard count. As per ES-12539, this is replaced with the {@link IndexShardCount} class that writes and loads only the index's
```
Likewise, no real need for a link to the task here, we can find that out from Git history if we really need it but most code readers will not care. If we did this for every issue we'd have 100k+ links to closed tickets in the source code, essentially none of which are valuable to readers of the code as it stands today.

DaveCTurner reviewed (commented):
One request about error checking during parsing and another few nits but otherwise this looks good.

joshua-adams-1 committed:
Comments

Addresses nit comments and adds extra protection against nested setting
objects

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

DaveCTurner reviewed (approved):
LGTM great stuff

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format

joshua-adams-1 committed:
Merge branch 'main' into new-index-metadata-format
